modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/toolbarbuttons/LinkUtils.js

```diff
@@ -0,0 +1,290 @@
+			}
+			else {
+				const style = new CKEDITOR.style({
+					alwaysRemoveElement: 1,
+					element: 'a',
+					type: CKEDITOR.STYLE_INLINE,
+				});
+
+				// 'removeStyle()' removes the style from the editor's current selection.
+				//  We need to force the selection to be the whole link element
+				//  to remove it properly.
+
+				const selection = editor.getSelection();
+				selection.selectElement(selection.getStartElement());
+
+				editor.removeStyle(style);
+			}
+		}
+
+		/**
+		 * Updates the href of an already existing link.
+		 *
+		 * @instance
+		 * @memberof CKEDITOR.Link
+		 * @method update
+		 * @param {CKEDITOR.dom.element} link The link element which href should be removed.
+		 * @param {Object|String} attrs The attributes to update or remove. Attributes with null values will be removed.
+		 * @param {Object} modifySelection A config object with an advance attribute to indicate if the selection should be moved after the link creation.
+		 */
+		update(attrs, link, modifySelection) {
+			const instance = this;
+
+			link = link || this.getFromSelection();
+
+			if (typeof attrs === 'string') {
+				const uri = instance._getCompleteURI(attrs);
+
+				link.setAttributes({
+					'data-cke-saved-href': uri,
+					href: uri,
+				});
+			}
+			else if (typeof attrs === 'object') {
+				const removeAttrs = [];
+
+				const setAttrs = {};
+
+				Object.keys(attrs).forEach((key) => {
+					if (attrs[key] === null) {
+						if (key === 'href') {
+							removeAttrs.push('data-cke-saved-href');
+						}
+
+						removeAttrs.push(key);
+					}
+					else {
+						if (key === 'href') {
+							const uri = instance._getCompleteURI(attrs[key]);
+
+							setAttrs['data-cke-saved-href'] = uri;
+							setAttrs[key] = uri;
+						}
+						else {
+							setAttrs[key] = attrs[key];
+						}
+					}
+				});
+
+				link.removeAttributes(removeAttrs);
+				link.setAttributes(setAttrs);
+			}
+
+			if (modifySelection && modifySelection.advance) {
+				this.advanceSelection(link);
+			}
+		}
+
+		/**
+		 * Checks if the URI begins with a '#' symbol to determine if it's an on page bookmark.
+		 * If it doesn't, it then checks if the URI has an '@' symbol. If it does and the URI
+		 * looks like an email and doesn't have 'mailto:', 'mailto:' is added to the URI.
+		 * If it doesn't and the URI doesn't have a scheme, the default 'http' scheme with
+		 * hierarchical path '//' is added to the URI.
+		 *
+		 * @instance
+		 * @memberof CKEDITOR.Link
+		 * @method _getCompleteURI
+		 * @param {String} URI The URI of the link.
+		 * @protected
+		 * @return {String} The URI updated with the protocol.
+		 */
+		_getCompleteURI(URI) {
+			if (REGEX_BOOKMARK_SCHEME.test(URI)) {
+				return URI;
+			}
+			else if (REGEX_EMAIL_SCHEME.test(URI)) {
+				URI = 'mailto:' + URI;
+			}
+			else if (!REGEX_URI_SCHEME.test(URI)) {
+				URI = this.appendProtocol ? 'http://' + URI : URI;
+			}
+
+			return URI;
+		}
+	}
+
+	Liferay.LinkUtilsCKEditor = Link;
```

Comment:
We should avoid adding to `Liferay`. Is there maybe an API to extend `CKEDITOR.tools`? Is there some other way ckeditor would use a general purpose utility?